apps/client/src/features/rundown/RundownEntry.tsx

```diff
@@ -95,6 +95,12 @@ export default function RundownEntry(props: RundownEntryProps) {
           addEvent({ type: SupportedEvent.Block }, { after: data.id });
           break;
         }
+        case 'swap': {
+          if (typeof payload === 'number') {
+            swapEvents({ from: payload, to: eventIndex });
```

Comment:
I think this might be the source of a bug.

With your current implementation
- Clear the rundown by Clicking the `Events...` button and choosing `Delete All`
- Add an event, we will call it A
- Add a second event, we will call it B

If we test the swap now between A and B everything works correctly

Now navigate again to the `Event...` button and select `Add delay at start` and retry the operation
From what I could see here, the swap fails

In the rundown, the eventIndex keeps track of the order of the entries of type Event.
In our previous example we would have
- delay (array index 0)
- A (array index 1, event index 0)
- B (array index 2, event index 1)

In short, we cannot use eventIndex as a signifier for the items position in the array

Reply:
I didn't test with delays! 
Okay, so we should use `id`s then. Do you think the backend and frontend code will be very similar since we'll be doing the same operation?

Reply:
Yes, I think the backend and frontend code will do exactly the same, to the exception that
- the backend needs to mutate both the `DataProvider` and the `delayedRundown`
- the `delayedRundown` also needs to move the delay field

Reply:
>  - the backend needs to mutate both the DataProvider and the delayedRundown
>  - the delayedRundown also needs to move the delay field

Can you please explain these two bullet points?

Reply:
Absolutely

###  the backend needs to mutate both the DataProvider and the delayedRundown
In the backend we essential keep the rundown duplicated.
Once inside the DataProvider, as is in the persistent store
and again in the delayedRundown, with some derived data

We need to remember to do both, and we use the functions like `cachedEdit` to achieve this
- [updating the data provider ](https://github.com/cpvalente/ontime/blob/master/apps/server/src/services/rundown-service/delayedRundown.utils.ts#L85)
- [updating the cached data](https://github.com/cpvalente/ontime/blob/master/apps/server/src/services/rundown-service/delayedRundown.utils.ts#L100)

### the delayedRundown also needs to move the delay field
the idea of the delayedRundown object is that it contains derived data from the original rundown. That includes a `delay` field which is important for this feature
we just need to make sure that the delay data gets swapped as we discussed [here](https://github.com/cpvalente/ontime/pull/446#discussion_r1286379502)